lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -12,549 +12,983 @@
+      return AugmentedLiteralTypeAttr::get(&getContext(), dict);
+  } else if (classBase == "DeletedType") {
+    if (auto name = dict.getAs<StringAttr>("name"))
+      return AugmentedDeletedTypeAttr::get(&getContext(), dict);
+  } else {
+    emitCircuitError() << "has an invalid AugmentedType";
+  }
+  return None;
 }
 
-/// Process all other operations.  This will throw an error if the operation
-/// contains any annotations that indicates that this should be included in an
-/// interface.  Otherwise, this is a valid nop.
-void GrandCentralVisitor::visitUnhandledDecl(Operation *op) {
-  AnnotationSet annotations(op);
-  auto anno = annotations.getAnnotation(
-      "sifive.enterprise.grandcentral.AugmentedGroundType");
-  if (!anno)
-    return;
-  auto diag =
-      op->emitOpError()
-      << "is marked as a an interface element, but this op or its ports are "
-         "not supposed to be interface elements (Are your annotations "
-         "malformed? Is this a missing feature that should be supported?)";
-  diag.attachNote()
-      << "this annotation marked the op as an interface element: '" << anno
-      << "'";
-  failed = true;
+bool GrandCentralPass::traverseField(Attribute field, IntegerAttr id,
+                                     Twine path) {
+  return TypeSwitch<Attribute, bool>(field)
+      .Case<AugmentedGroundTypeAttr>([&](auto ground) {
+        Value leafValue = leafMap.lookup(ground.getID());
+        assert(leafValue && "leafValue not found");
+
+        auto builder = OpBuilder::atBlockEnd(
+            companionIDMap.lookup(id).mapping.getBodyBlock());
+
+        auto srcPaths =
+            instancePaths->getAbsolutePaths(getEnclosingModule(leafValue));
+        assert(srcPaths.size() == 1 &&
+               "Unable to handle multiply instantiated companions");
+        SmallString<0> srcRef;
+        for (auto path : srcPaths[0])
+          srcRef.append((path.name() + ".").str());
+
+        auto uloc = builder.getUnknownLoc();
+        if (auto blockArg = leafValue.dyn_cast<BlockArgument>()) {
+          FModuleOp module =
+              cast<FModuleOp>(blockArg.getOwner()->getParentOp());
+          builder.create<sv::VerbatimOp>(
+              uloc, "assign " + path + " = " + srcRef +
+                        module.portNames()[blockArg.getArgNumber()]
+                            .cast<StringAttr>()
+                            .getValue() +
+                        ";");
+        } else
+          builder.create<sv::VerbatimOp>(uloc, "assign " + path + " = " +
+                                                   srcRef +
+                                                   leafValue.getDefiningOp()
+                                                       ->getAttr("name")
+                                                       .cast<StringAttr>()
+                                                       .getValue() +
+                                                   ";");
```

Comment:
Can we do anything about the unfortunate formating here? If not, its ok. 

```c++
      StringRef name;
      if (auto blockArg = leafValue.dyn_cast<BlockArgument>()) {
        auto module = cast<FModuleLike>(blockArg.getOwner()->getParentOp());
        name = module.portNames()[blockArg.getArgNumber()]

      } else {
        auto op = leafValue.getDefiningOp();
        leafName = op->getAttr("name").cast<StringAttr>().getValue()
      }
      builder.create<sv::VerbatimOp>(uloc, "assign " + path + " = " + srcRef +
                                               leafName ";");

```

Reply:
Sometimes clang format is trying to be waaaaay too clever for my taste.